draft-ietf-ipsecme-ikev2-multiple-ke-00.xml

```diff
@@ -520,21 +550,103 @@ Initiator                             Responder
+                according to its local policy.
                 </t>
 
-                <t> If the initiator includes any Additional Key Exchanges transform types into SA payload,
+                <t> The responder performs negotiation using standard IKEv2 procedure described in Section 3.3 of <xref target="RFC7296"/>.
+                However, for the Additional Key Exchange types the responder's choice MUST NOT contain equal transform IDs (apart from NONE), 
+                and the ID selected for Transform Type 4 MUST NOT appear in any of Additional Key Exchange transforms.
+                In other words, all selected key exchange methods must be different.
+                </t>
+
+                <t> If the responder selected NONE for some Additional Key Exchange types (provided they were proposed by the initiator), 
+                then the corresponding IKE_INTERMEDIATE exchanges should not take place. The IKE_INTERMEDIATE exchanges
+                MUST only be performed for Additional Key Exchange types containing non-NONE responders choices.
+                It means that if the initiator includes NONE in all Additional Key Exchange transforms and the
+                responder selects this value for all of them, then no IKE_INTERMEDIATE exchanges that would have
+                perform additional key exchanges will take place (note that they still may take place for other purposes).
+                </t>
+
+                <t> Below is an example of the SA payload in the initiator's IKE_SA_INIT request message.
+                Here we use an abbreviation AKE1, AKE 2 etc. to denote Additional Key Exchange 1, 
+                Additional Key Exchange 2 etc. transforms, that this document defines, 
+                and an abbreviation KE for the Key Exchange transform,
+                that this document renames from the Diffie-Hellman Group transform.
+                We also use not yet defined Transform IDs FRODOKEM_976_AES, SIKE_P434
+                and KYBER512 to denote some of popular post-quantum key exchange methods.
+                </t>
+
+                <figure><artwork align="center" ><![CDATA[
+SA Payload
+   |
+   +--- Proposal #1 ( Proto ID = IKE(1), SPI size = 8,
+         |            9 transforms,      SPI = 0x35a1d6f22564f89d )
+         |
+         +-- Transform ENCR ( ID = ENCR_AES_GCM_16 )
+         |     +-- Attribute ( Key Length = 256 )
+         |
+         +-- Transform KE ( ID = 4096-bit MODP Group )
+         |
+         +-- Transform PRF ( ID = PRF_HMAC_SHA2_256 )
+         |
+         +-- Transform AKE2 ( ID = FRODOKEM_976_AES )
+         |
+         +-- Transform AKE2 ( ID = SIKE_P434 )
+         |
+         +-- Transform AKE3 ( ID = FRODOKEM_976_AES )
+         |
+         +-- Transform AKE3 ( ID = SIKE_P434 )
+         |
+         +-- Transform AKE5 ( ID = KYBER512 )
+         |
+         +-- Transform AKE5 ( ID = NONE )
+                ]]></artwork></figure>
```

Comment:
> I think we should avoid specific algorithm names (i.e. `FRODOKEM_976_AES`, `SIKE_P434` and `KYBER512`). They are still NIST candidates and we don't want to be seen to "favour" these algorithms, or influencing selection, etc. Perhaps we can use some abstract name, e.g.: NIST_PQ_KEM_1, NIST_PQ_KEM_2, etc.
> 
> Or if anyone can come up with a better name, happy to use that too.

OK, makes sense.

Reply:
> The transform `AKE1` is not listed and I guess this is on-purpose?
> 
> Do you envisage that we will have a standard list of what transform IDs that will be included in `AKE1`, `AKE2`, etc?
> 
> Just curious, in practice, what circumstances we will skip `AKE1`. Usually, the transform type `AKEi` would be preferred than `AKEj` where `i < j`.

Yes, on purpose to illustrate that there is no requirement in the draft that proposed AKEs must start from AKE1, followed by AKE2 (if any) etc. It's perfectly valid to propose only, say, AKE3 and AKE6. I can imagine that some implementers may associate specific AKEs with specific PQ KEMs, e.g. AKE1 contains lattice-based KEMs, AKE2 - code based, AKE3 - isogeny based. In this case depending on local policy (say, security officer decided to only rely on code-based KEMs), initiator may only propose AKE2. I don't know how practical it is, just don't want to add restrictions to prevent this scenario.

Reply:
Ok, yep that makes sense and I have a follow-on question on this subject and it might be related to what Paul asked on the mailing list. But let's not hold this up, I'll follow up with you separately on this question.

I'll merge this PR and I'll rename some of the specific PQ algorithms to generic names, i.e. FRODOKEM_976_AES to PQ_KEM_X.